pkg/datasource/data.go

```diff
@@ -0,0 +1,419 @@
+		}
+		newFields = append(newFields, nf)
+	}
+
+	var err error
+	for i, f := range newFields {
+		f.FullName, err = resolveNames(uint32(i), newFields, uint32(parentOffset))
+		if err != nil {
+			return nil, fmt.Errorf("resolvong full fieldnames: %w", err)
+		}
+	}
+
+	ds.fields = append(ds.fields, newFields...)
+
+	for _, f := range newFields {
+		ds.fieldMap[f.Name] = f
+	}
+
+	ds.payloadCount++
+
+	return &fieldAccessor{ds: ds, f: &field{
+		PayloadIndex: idx,
+		Size:         size,
+	}}, nil
+}
+
+func (ds *dataSource) AddField(name string, opts ...FieldOption) (FieldAccessor, error) {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if _, ok := ds.fieldMap[name]; ok {
+		return nil, fmt.Errorf("field %q already exists", name)
+	}
+
+	nf := &field{
+		Name:     name,
+		FullName: name,
+		Index:    uint32(len(ds.fields)),
+		Kind:     api.Kind_Invalid,
+	}
+	for _, opt := range opts {
+		opt(nf)
+	}
+
+	// Reserve new payload for non-empty fields
+	if !FieldFlagEmpty.In(nf.Flags) {
+		nf.PayloadIndex = ds.payloadCount
+		ds.payloadCount++
+	}
+
+	ds.fields = append(ds.fields, nf)
+	ds.fieldMap[nf.FullName] = nf
+	return &fieldAccessor{ds: ds, f: nf}, nil
+}
+
+func (ds *dataSource) GetField(name string) FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	f, ok := ds.fieldMap[name]
+	if !ok {
+		return nil
+	}
+	return &fieldAccessor{ds: ds, f: f}
+}
+
+func (ds *dataSource) GetFieldsWithTag(tag ...string) []FieldAccessor {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	res := make([]FieldAccessor, 0)
+	for _, f := range ds.fields {
+		for _, t := range tag {
+			if slices.Contains(f.Tags, t) {
+				res = append(res, &fieldAccessor{ds: ds, f: f})
+			}
+		}
+	}
+	return res
+}
+
+func (ds *dataSource) Subscribe(fn DataFunc, priority int) {
+	if fn == nil {
+		return
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	ds.subscriptions = append(ds.subscriptions, &subscription{
+		priority: priority,
+		fn:       fn,
+	})
+	sort.SliceStable(ds.subscriptions, func(i, j int) bool {
+		return ds.subscriptions[i].priority < ds.subscriptions[j].priority
+	})
+}
+
+func (ds *dataSource) EmitAndRelease(d Data) error {
+	defer ds.dPool.Put(d)
+	for _, sub := range ds.subscriptions {
+		err := sub.fn(ds, d)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ds *dataSource) Release(d Data) {
+	ds.dPool.Put(d)
+}
+
+func (ds *dataSource) ReportLostData(ctr uint64) {
+	// TODO
+}
+
+func (ds *dataSource) IsRequestedField(fieldName string) bool {
+	return true
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+	return ds.requestedFields[fieldName]
+}
+
+func (ds *dataSource) Dump(xd Data, wr io.Writer) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	d := xd.(*data)
+	for _, f := range ds.fields {
+		if f.Offs+f.Size > uint32(len(d.Payload[f.PayloadIndex])) {
+			fmt.Fprintf(wr, "%s (%d): ! invalid size\n", f.Name, f.Size)
```

Comment:
It's just a debugging function to dump `Data` to a writer - in this case, the error would be printed to the dump. This should usually not happen, because the checks should happen in other places, but especially for debugging purposes I left that in.